### DISCUSSION TITLE
polling basestation network

### DISCUSSION THREAD

**Discussion by woudie-swap** [semuconsulting/pygnssutils#49] [Q&A]
Hi,

I was interested in using this package to create a small utility to poll and monitor my network of base stations primarily available through ntrip connections without my rover receiver being online yet. As far as I understand, that is something I can do with this package, but I am a bit confused about how to about about that right now - can some tips be provided? Any advice or insight on how to go about this would be greatly appreciated!

**Comment by semuadmin** [comment]
Hi @woudie-swap 

So if I understand correctly, you want to monitor the RTCM output from one or more NTRIP casters? That is certainly something you can do with the [gnssntripclient](https://github.com/semuconsulting/pygnssutils#gnssntripclient) CLI utility or its underlying [GNSSNTRIPClient](https://github.com/semuconsulting/pygnssutils/blob/main/src/pygnssutils/gnssntripclient.py) Python class.

If you want to monitor multiple casters simultaneously, you'd need to run a separate instance for each caster.

e.g.

```shell
> gnssntripclient --server rtk2go.com --mountpoint UFOSRUS  --ntripuser myuser --ntrippassword mypassword --ggamode 1 --ggainterval 60 --reflat 37.23 --reflon -115.81 
```
```
2022-06-03 11:55:10.305870: <RTCM(1077, DF002=1077, DF003=0, GNSSEpoch=471328000, DF393=1, ...
```

I recommend you refer to the [README](https://github.com/semuconsulting/pygnssutils#gnssntripclient) and [Sphinx API documentation](https://www.semuconsulting.com/pygnssutils/pygnssutils.html#module-pygnssutils.gnssntripclient) in the first instance, and get back to me with any specific queries.

If you're new to NTRIP, there is an introductory overview [here](https://www.semuconsulting.com/gnsswiki/#DGPS).

What is your ultimate objective here? What are you hoping to achieve by monitoring this data?

**Reply by woudie-swap** [reply]
Hi @semuadmin,

Ah yes, sorry I should have clarified - I played around with the CLI utility and examples to stream various RTCM messages from various casters. 

I am trying to build a small utility to corrections bridge utility that allows me to monitor the realtime position and availability of corrections data from various sources (base stations or VRS via ntrip, lora, uhf, etc), perform quality checks and forward the best available to my receiver

**Reply by semuadmin** [reply]
Unfortunately I don't think `pygnssutils` will help with this.

IP, LORA and UHF simply refer to different data transmission mechanisms - what's relevant here is the RTK protocol that's actually being transmitted, which is presumably the same in each case i.e. RTCM.

The RTCM data from the various sources you mention is combined with live observations from the receiver itself and processed within the receiver's firmware to arrive at a 'corrected' navigation solution. If you wanted to compare what those 'corrected' navigation solutions look like _before_ they are processed by the receiver, you'd need to use a library capable of performing the equivalent RTK calculations - i.e. something like RTKLIB. This is not what `pygnssutils` is designed to do.

**Reply by woudie-swap** [reply]
Hi @semuadmin 

Yes, concerning evaluating the correction data, I already have a different solution in mind involved rtklibs `rtkrcv`. 

I am interested in using `pygnssutils` to help me gather info about GPS' that are the corrections sources (position coordinates & availability primarily) similar to what is available in a ntrip source table. I am only interested in supporting RTCM messages. I tried looking at the RTCM messages and saw that `1005` or `1006` are able to provide the info I am looking for, but I didn't see those being received when I was testing - which is what led to the confusion.

Would I be able to use `pygnssutils` in this respect?

**Reply by semuadmin** [reply]
So I would refer you back to the original answer - the gnssntripclient CLI utility can parse RTCM data from any NTRIP source  - for example, here's the log output from a local basestation which includes the RTCM 1006 message (_if the base station isn't supplying a 1005 or 1006 message, check that it supports RTCM3 and whether it requires a GGA position sentence from your client_).

```shell
gnssntripclient -S euref-ip.net -M DARE00GBR0 --ntripuser semuadmin --ntrippassword **********
```

```
...
2023-12-15 17:46:44.681728: <RTCM(1006, DF002=1006, DF003=25, DF021=0, DF022=1, DF023=1, DF024=0, DF141=0, DF025=3811965.6701, DF142=1, DF001_1=0, DF026=-175800.07080000002, DF364=1, DF027=5093615.5606, DF028=0.1)>
...
```

DF003 = Reference station ID (not always populated)
DF025, 26 & 27 = ECEF X,Y,Z coordinates of the reference station
DF028 = Antenna height above survey marker (1006 messages only - omitted in 1005)

If you want to use these values programmatically in a Python script, you can use the [`pyrtcm`](https://github.com/semuconsulting/pyrtcm) or [`pyubx2`](https://github.com/semuconsulting/pyubx2) libraries to parse and decode any binary RTCM data stream (e.g. binary log file or serial stream):

```python
from pyubx2 import UBXReader, ecef2llh, RTCM3_PROTOCOL

with open("rtcm.bin", "rb") as stream:
    ubr = UBXReader(stream, protfilter=RTCM3_PROTOCOL)
    for raw, parsed in ubr:
        if parsed.identity in ("1005", "1006"):
            arp = parsed.DF003
            ecef_x, ecef_y, ecef_z = parsed.DF025, parsed.DF026, parsed.DF027
            lat, lon, alt = ecef2llh(ecef_x, ecef_y, ecef_z)
            print(f"ARP {arp}: lat {lat}, lon {lon}, alt {alt}")
            break
```

```
ARP 25: lat 53.344802702956386, lon -2.640493393705157, alt 88.51139419065267
```

(_the above example uses a `pyubx2` utility function to convert ECEF to LLH for ease of use; for RTK purposes you'd probably want to use the ECEF coordinates_)

Hope this helps
